src/frontend/js/classes/table-of-content.js

```diff
@@ -0,0 +1,289 @@
+/**
+ * Generate dynamic table of content
+ */
+export default class TableOfContent {
+  /**
+   * Initialize table of content
+   *
+   * @param {string} tagSelector - selector for tags to observe
+   * @param {string} tocParentElement - selector for table of content wrapper
+   * @param {string|null} fixedHeaderSelector - selector for fixed site header
+   */
+  constructor({ tagSelector, tocParentElement, fixedHeaderSelector = null }) {
+    /**
+     * Array of tags to observe
+     */
+    this.tags = [];
+    this.tagsSectionsMap = [];
+
+    /**
+     * Selector for tags to observe
+     */
+    this.tagSelector = tagSelector || 'h2,h3,h4';
+
+    /**
+     * Selector for table of content wrapper
+     */
+    this.tocParentElement = tocParentElement;
+
+    /**
+     * Selector for getting real height of the fixed header to detect active section correctly
+     * @type {null}
+     */
+    this.fixedHeaderSelector = fixedHeaderSelector;
+
+    this.CSS = {
+      tocContainer: 'table-of-content',
+      tocHeader: 'table-of-content__header',
+      tocElement: 'table-of-content__list',
+      tocElementItem: 'table-of-content__list-item',
+      tocElementItemActive: 'table-of-content__list-item--active',
+      tocElementItemIndent: number => `table-of-content__list-item--indent-${number}x`,
+    };
+
+    this.init();
+  }
+
+  /**
+   * Initialize table of content
+   */
+  init() {
+    this.tags = this.getSectionTagsOnThePage();
+
+    /**
+     * Check if no tags found then table of content is not needed
+     */
+    if (this.tags.length === 0) {
+      return;
+    }
+
+    /**
+     * Initialize table of content element
+     */
+    this.createTableOfContent();
+    this.addTableOfContent();
+
+    /**
+     * Calculate boundings for each tag and watch active section
+     */
+    this.calculateBoundings();
+    this.watchActiveSection();
+  }
+
+  /**
+   * Find all section tags on the page
+   *
+   * @return {HTMLElement[]}
+   */
+  getSectionTagsOnThePage() {
+    return Array.from(document.querySelectorAll(this.tagSelector));
+  }
+
+  /**
+   * Calculate top line position for each tag
+   */
+  calculateBoundings() {
+    this.tagsSectionsMap = this.tags.map((tag) => {
+      const rect = tag.getBoundingClientRect();
+      const top = rect.top + window.scrollY;
```

Comment:
you can add the third parameter like "offset" to resolve the problem like that

![image](https://user-images.githubusercontent.com/3684889/178775154-80c18ac1-07d1-4a8e-8afd-3f48a9b9c1b2.png)